pandas/tests/indexes/test_setops.py

```diff
@@ -501,6 +501,81 @@ def check_intersection_commutative(left, right):
     assert idx.intersection(idx_non_unique).is_unique
 
 
+@pytest.mark.parametrize(
+    "cls",
+    [
+        Int64Index,
+        Float64Index,
+        DatetimeIndex,
+        CategoricalIndex,
+        TimedeltaIndex,
+        lambda x: Index(x, dtype=object),
+    ],
+)
+def test_union_duplicate_index_subsets_of_each_other(cls):
+    # GH#31326
+    a = cls([1, 2, 2, 3])
+    b = cls([3, 3, 4])
+    expected = cls([1, 2, 2, 3, 3, 4])
+    if cls is CategoricalIndex:
+        expected = Index([1, 2, 2, 3, 3, 4])
```

Comment:
IIUC it would still be CategoricalIndex if a and b had a dtype `CategoricalDtype(categories=[1, 2, 3, 4])`?

Reply:
Does not seem so. Is cast to Index for all cases, also for parts I did not touch